### PR TITLE
create leaderboard

### PR DESCRIPTION
**Pull request by pbochis**

### CODE REVIEW

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/repository/ResultRepository.java [hunk 1]
```diff
@@ -28,4 +28,12 @@
             "JOIN challenge.challengeTemplate template " +
             "WHERE template.id = :templateId ")
     List<Result> findAllByTemplate(@Param("template") UUID templateId);
+
+    @Query(" SELECT result, count(taskResult), max(taskResult.endTime) FROM Result result " +
```
Comment (victorbalan) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/repository/ResultRepository.java]: I think you can use 
```
SELECT new cod.uno.platform.server.kernel.dto.LeaderBoardEntry(
   new cod.uno.platform.dto.UserShowDto(result.user), count(taskResult), max(taskResult.endTime)) FROM

```

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/service/TaskResultService.java [hunk 2]
```diff
@@ -36,18 +36,6 @@ public TaskResultService(TaskResultRepository taskResultRepository,
     public void startTask(UUID resultId, UUID taskId) throws IllegalArgumentException {
         Result result = resultRepository.findOneWithChallenge(resultId);
         Task task = taskRepository.findOne(taskId);
-        List<Task> tasks = result.getChallenge().getChallengeTemplate().getTasks();
```
Comment (victorbalan) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/service/TaskResultService.java]: remove this..i already have a pr with the fix

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/dto/user/UserShowDto.java [hunk 3]
```diff
@@ -8,6 +8,7 @@
 public class UserShowDto {
     private UUID id;
     private String username;
+    private String email;
```
Comment (Exordian) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/dto/user/UserShowDto.java]: no

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/dto/challenge/LeaderboardEntry.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+package uno.cod.platform.server.core.dto.challenge;
+
+import uno.cod.platform.server.core.dto.user.UserShowDto;
+
+import java.time.ZonedDateTime;
+
+public class LeaderboardEntry {
```
Comment (Exordian) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/dto/challenge/LeaderboardEntry.java]: why is this the only class that does not end with Dto in the dto package?

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/repository/ResultRepository.java [hunk 5]
```diff
@@ -28,4 +28,12 @@
...
+            "JOIN result.challenge challenge " +
+            "JOIN result.taskResults taskResult " +
+            "WHERE challenge.id =:challengeId and taskResult.green=true " +
+            "GROUP BY result.user " +
+            "ORDER BY COUNT(taskResult) DESC, MAX(taskResult.endTime) ASC ")
```
Comment (victorbalan) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/repository/ResultRepository.java]: you count all taskResults? `and taskResult.green=true ` this does not mean you select only green taskResults..it means you select results with green taskResults

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/service/ResultService.java [hunk 6]
```diff
@@ -49,4 +53,18 @@ public ResultShowDto findOne(UUID id){
     public ResultShowDto findOneByUserAndChallenge(UUID userId, UUID challengeId){
         return ResultMapper.map(repository.findOneByUserAndChallenge(userId, challengeId));
     }
+
+    public List<LeaderboardEntry> getLeaderboard(UUID challengeId){
```
Comment (victorbalan) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/service/ResultService.java]: there is no unit test for this method

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/dto/challenge/LeaderboardEntry.java (hunk 4 above)
Comment (victorbalan) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/dto/challenge/LeaderboardEntry.java]: please rename this to `LeaderboardEntryDto`

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/domain/Result.java [hunk 7]
```diff
@@ -21,6 +21,7 @@
 
     private ZonedDateTime finished;
 
+    @OrderBy("startTime")
```
Comment (victorbalan) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/domain/Result.java]: why order by startTime?

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/repository/ResultRepository.java (hunk 5 above)
Comment (pbochis) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/repository/ResultRepository.java]: I'll fix the query.

File: platform-server-kernel/src/main/java/uno/cod/platform/server/core/service/TaskResultService.java (hunk 2 above)
Comment (pbochis) [inline platform-server-kernel/src/main/java/uno/cod/platform/server/core/service/TaskResultService.java]: i will rebase.

**Pull request merged**